protocol.go

```diff
@@ -0,0 +1,426 @@
+
+	if !header.Unordered && !p.trackRead(header.Sequence) {
+		return p.addr, nil, nil
+	}
+
+	p.trackUnacked()
+
+	if header.Empty {
+		return p.addr, nil, nil
+	}
+
+	if p.ph != nil {
+		p.ph(p.addr, header.Sequence, buf)
+	}
+
+	//log.Printf("%s: recv    (seq=%05d) (ack=%05d) (ack_bits=%032b) (size=%d) (reliable=%t)", p.conn.LocalAddr(), header.Sequence, header.ACK, header.ACKBits, len(buf), !header.Unordered)
+
+	return p.addr, p.writeAcksIfNecessary(), nil
+}
+
+func (p *Protocol) createAckIfNecessary() (header PacketHeader, needed bool) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	lui := p.lui
+
+	for i := uint16(0); i < ACKBitsetSize; i++ {
+		if p.rq[(lui+i)%uint16(len(p.rq))] != uint32(lui+i) {
+			return header, needed
+		}
+	}
+
+	lui += ACKBitsetSize
+	p.lui = lui
+	p.ls = time.Now()
+
+	p.waitUntilReaderAvailable()
+
+	header.Sequence, header.ACK = p.nextWriteIndex(), lui-1
+	header.ACKBits = p.prepareAckBits(header.ACK)
+	header.Empty = true
+
+	needed = !p.die
+
+	return header, needed
+}
+
+func (p *Protocol) writeAcksIfNecessary() []byte {
+	for {
+		header, needed := p.createAckIfNecessary()
+		if !needed {
+			return nil
+		}
+
+		//log.Printf("%s: ack     (seq=%05d) (ack=%05d) (ack_bits=%032b)", p.conn.LocalAddr(), header.Sequence, header.ACK, header.ACKBits)
+
+		return p.write(header, nil)
+	}
+}
+
+func (p *Protocol) readAckBits(ack uint16, ackBits uint32) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for idx := uint16(0); idx < ACKBitsetSize; idx, ackBits = idx+1, ackBits>>1 {
+		if ackBits&1 == 0 {
+			continue
+		}
+
+		i := (ack - idx) % uint16(len(p.wq))
+		if p.wq[i] != uint32(ack-idx) || p.wqe[i].acked {
+			continue
+		}
+
+		if p.wqe[i].buf != nil {
+			p.pool.Put(p.wqe[i].buf)
+		}
+
+		p.wqe[i].buf = nil
+		p.wqe[i].acked = true
+	}
+}
+
+func (p *Protocol) trackRead(idx uint16) bool {
+	i := idx % uint16(len(p.rq))
+
+	if p.rq[i] == uint32(idx) { // duplicate packet
+		return false
+	}
+
+	if seq.GT(idx+1, p.ri) {
+		p.clearReads(p.ri, idx)
+		p.ri = idx + 1
+	}
+
+	p.rq[i] = uint32(idx)
+
+	return true
+}
+
+func (p *Protocol) clearReads(start, end uint16) {
+	count, size := end-start+1, uint16(len(p.rq))
+
+	if count >= size {
+		emptyBufferIndices(p.rq)
+		return
+	}
+
+	first := p.rq[start%size:]
+	length := uint16(len(first))
+
+	if count <= length {
+		emptyBufferIndices(first[:count])
+		return
+	}
+
+	second := p.rq[:count-length]
+
+	emptyBufferIndices(first)
+	emptyBufferIndices(second)
+}
+
+func (p *Protocol) trackAcked(ack uint16) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	lui := p.lui
+
+	for lui <= ack {
+		if p.rq[lui%uint16(len(p.rq))] != uint32(lui) {
+			break
+		}
+		lui++
+	}
+
+	p.lui = lui
+	p.ls = time.Now()
+}
+
+func (p *Protocol) trackUnacked() {
+	oui := p.oui
+
+	for {
+		i := oui % uint16(len(p.wq))
+		if p.wq[i] != uint32(oui) || !p.wqe[i].acked {
+			break
+		}
+		oui++
+	}
+	p.oui = oui
+
+	p.ouc.Broadcast()
+}
+
+func (p *Protocol) close() bool {
```

Comment:
If two goroutines call close() at once, `close(p.exit)` can be called twice and cause a panic.

Reply:
Yes, exactly.